File: examples/cpp/HFR/hfr_small_preview.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <optional>

#include "depthai/depthai.hpp"

constexpr std::pair<int, int> SIZE = {1280, 720};
constexpr int FPS = 480;

int main() {
    dai::Pipeline pipeline;

    auto device = pipeline.getDefaultDevice();
    auto platform = device->getPlatform();
    if(platform != dai::Platform::RVC4) {
        std::cerr << "This example is only supported on IMX586 and Luxonis OS 1.20.5 or higher\n" << std::flush;
        return 0;
    }

    // Exit cleanly if the selected HFR mode is not advertised by CAM_A.
    bool supportsRequestedFps = false;
    for(const auto& cameraFeature : device->getConnectedCameraFeatures()) {
        if(cameraFeature.socket != dai::CameraBoardSocket::CAM_A) continue;
        for(const auto& config : cameraFeature.configs) {
            if(config.width == SIZE.first && config.height == SIZE.second && config.maxFps >= static_cast<float>(FPS)) {
                supportsRequestedFps = true;
                break;
            }
        }
        break;
    }
    if(!supportsRequestedFps) {
        std::cerr << "This example is only supported on IMX586 and Luxonis OS 1.20.5 or higher\n" << std::flush;
        return 0;
    }

    auto cam = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::AUTO, std::nullopt, static_cast<float>(FPS));
    auto benchmarkIn = pipeline.create<dai::node::BenchmarkIn>();
    benchmarkIn->setRunOnHost(true);
    benchmarkIn->sendReportEveryNMessages(FPS);

    auto* output = cam->requestOutput(std::make_pair(250U, 250U));
    output->link(benchmarkIn->input);

    auto outputQueue = output->createOutputQueue();

    pipeline.start();

    while(pipeline.isRunning()) {
        auto imgFrame = outputQueue->get<dai::ImgFrame>();
        if(imgFrame == nullptr) continue;
        cv::imshow("frame", imgFrame->getCvFrame());
        if(cv::waitKey(1) == 'q') {
            break;
        }
    }

    return 0;
}

File: examples/python/HFR/hfr_small_preview.py
#!/usr/bin/env python3
import depthai as dai
import sys
import time
import cv2

SIZE = (1280, 720)
FPS = 480

# SIZE = (1920, 1080)
# FPS = 240

with dai.Pipeline() as pipeline:
    device = pipeline.getDefaultDevice()
    platform = device.getPlatform()
    if platform != dai.Platform.RVC4:
        print("This example is only supported on IMX586 and Luxonis OS 1.20.5 or higher", file=sys.stderr)
        sys.exit(0)

    # Exit cleanly if the selected HFR mode is not advertised by CAM_A.
    supportsRequestedFps = False
    for cameraFeature in device.getConnectedCameraFeatures():
        if cameraFeature.socket != dai.CameraBoardSocket.CAM_A:
            continue
        for config in cameraFeature.configs:
            if config.width == SIZE[0] and config.height == SIZE[1] and config.maxFps >= FPS:
                supportsRequestedFps = True
                break
        break
    if not supportsRequestedFps:
        print("This example is only supported on IMX586 and Luxonis OS 1.20.5 or higher", file=sys.stderr)
        sys.exit(0)

    cam = pipeline.create(dai.node.Camera).build(sensorFps=FPS)
    benchmarkIn = pipeline.create(dai.node.BenchmarkIn)
    benchmarkIn.setRunOnHost(True)
    benchmarkIn.sendReportEveryNMessages(FPS)

    output = cam.requestOutput((250, 250))
    output.link(benchmarkIn.input)

    outputQueue = output.createOutputQueue()

    pipeline.start()
    while pipeline.isRunning():
        imgFrame = outputQueue.get()
        assert isinstance(imgFrame, dai.ImgFrame)
        cv2.imshow("frame", imgFrame.getCvFrame())
        cv2.waitKey(1)
